The dataset contains information on gun deaths in the US from 2012 to 2014. Each row in the dataset represents a single fatality. The columns contain demographic and other information about the victim. 

Here's an explanation of each column:
-- this is an identifier column, which contains the row number. It's common in CSV files to include a unique identifier for each row, but we can ignore it in this analysis.
year -- the year in which the fatality occurred.
month -- the month in which the fatality occurred.
intent -- the intent of the perpetrator of the crime. This can be Suicide, Accidental, NA, Homicide, or Undetermined.
police -- whether a police officer was involved with the shooting. Either 0 (false) or 1 (true).
sex -- the gender of the victim. Either M or F.
age -- the age of the victim.
race -- the race of the victim. Either Asian/Pacific Islander, Native American/Native Alaskan, Black, Hispanic, or White.
hispanic -- a code indicating the Hispanic origin of the victim.
place -- where the shooting occurred. Has several categories, which you're encouraged to explore on your own.
education -- educational status of the victim. Can be one of the following:
1 -- Less than High School
2 -- Graduated from High School or equivalent
3 -- Some College
4 -- At least graduated from College
5 -- Not available
In this project, we'll explore the dataset, and try to find patterns in the demographics of the victims. Our first step is to read the data in and take a look at it.

# Read the dataset in as a list using the csv module.
1. Import the csv module.
2. Open the file using the open() function.
3. Use the csv.reader() function to load the opened file.
4. Call list() on the result to get a list of all the data in the file.
5. Assign the result to the variable data.
6. Display the first 5 rows of data to verify everything.

In [21]:
import csv
import datetime
data = list(csv.reader(open('guns.csv','r')))
print(data[0:5])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education'], ['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]


# Removing Headers From A List Of Lists
1. Extract the first row of data, and assign it to the variable headers.
2. Remove the first row from data.
3. Display headers.
4. Display the first 5 rows of data to verify that you removed the header row properly.

In [18]:
headers = data[0]
headers = data[:1]
data = data[1:len(data)]
print("\n")
print("Headers", "\n","======" , "\n",headers)
print("\n")
print(data[:5])



Headers 
 [['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4']]


[['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', '2'], ['6', '2012', '02', 'Suicide', '0', 'M', '17', 'Native American/Native Alaskan', '100', 'Home', '1']]


# Counting Gun Deaths By Year
1. Use a list comprehension to extract the year column from data.
Because the year column is the second column in the data, you'll need to get the element at index 1 in each row.
Assign the result to the variable years.
2. Create an empty dictionary called year_counts.
3. Loop through each element in years.
If the element isn't a key in year_counts, create it, and set the value to 1.
If the element is a key in year_counts, increment the value by one.
4. Display year_counts to see how many gun deaths occur in each year.

In [3]:
years = [i[1] for i in data]
print(years[0:5])
year_counts = {}
for i in years:
    if i in year_counts:
        year_counts[i] = year_counts[i] + 1
    else:
        year_counts[i] = 1
        
print(year_counts)

['2012', '2012', '2012', '2012', '2012']
{'2013': 33636, '2014': 33599, '2012': 33563}


# Exploring Gun Deaths By Month And Year
1. Use a list comprehension to create a datetime.datetime object for each row. Assign the result to dates.
The year column in the second element in each row.
The month column is the third element in each row.
Make sure to convert year and month to integers using int().
Pass year, month, and day=1 into the datetime.datetime() function.
2. Display the first 5 rows in dates to verify everything worked.
3. Count up how many times each unique date occurs in dates. Assign the result to date_counts.
This follows a similar procedure to what we did in the last screen with year_counts.
4. Display date_counts.

In [4]:
#Exploring gun deaths by month, year
dates = []
for i in data:
    tmp = datetime.datetime(year=int(i[1]), month=int(i[2]), day=1)
    dates.append(tmp)
    
print(dates[:5])
print("\n")

date_counts = {}
for i in dates:
    if i in date_counts:
        date_counts[i] = date_counts[i] + 1
    else:
        date_counts[i] = 1

[datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0)]




In [10]:
date_counts

{datetime.datetime(2012, 1, 1, 0, 0): 2758,
 datetime.datetime(2012, 2, 1, 0, 0): 2357,
 datetime.datetime(2012, 3, 1, 0, 0): 2743,
 datetime.datetime(2012, 4, 1, 0, 0): 2795,
 datetime.datetime(2012, 5, 1, 0, 0): 2999,
 datetime.datetime(2012, 6, 1, 0, 0): 2826,
 datetime.datetime(2012, 7, 1, 0, 0): 3026,
 datetime.datetime(2012, 8, 1, 0, 0): 2954,
 datetime.datetime(2012, 9, 1, 0, 0): 2852,
 datetime.datetime(2012, 10, 1, 0, 0): 2733,
 datetime.datetime(2012, 11, 1, 0, 0): 2729,
 datetime.datetime(2012, 12, 1, 0, 0): 2791,
 datetime.datetime(2013, 1, 1, 0, 0): 2864,
 datetime.datetime(2013, 2, 1, 0, 0): 2375,
 datetime.datetime(2013, 3, 1, 0, 0): 2862,
 datetime.datetime(2013, 4, 1, 0, 0): 2798,
 datetime.datetime(2013, 5, 1, 0, 0): 2806,
 datetime.datetime(2013, 6, 1, 0, 0): 2920,
 datetime.datetime(2013, 7, 1, 0, 0): 3079,
 datetime.datetime(2013, 8, 1, 0, 0): 2859,
 datetime.datetime(2013, 9, 1, 0, 0): 2742,
 datetime.datetime(2013, 10, 1, 0, 0): 2808,
 datetime.datetime(2013, 11,

# Exploring Gun Deaths By Race And Sex
1. Count up how many times each item in the sex column occurs.
Assign the result to sex_counts.
2. Count up how many times each item in the race column occurs.
Assign the result to race_counts.
3. Display race_counts and sex_counts to verify your work, and see if you can spot any patterns.
4. Write a markdown cell detailing what you've learned so far, and what you think might need further examination.

In [5]:
#Exploring gun deaths by sex
sex_counts = {}
for i in data:
    if i[5] in sex_counts:
        sex_counts[i[5]] = sex_counts[i[5]] + 1
    else:
        sex_counts[i[5]] = 1
print(sex_counts)

{'M': 86349, 'F': 14449}


In [6]:
#Exploring gun deaths by race
race_counts = {}
for i in data:
    if i[7] in race_counts:
        race_counts[i[7]] = race_counts[i[7]] + 1
    else:
        race_counts[i[7]] = 1
print(race_counts)

{'White': 66237, 'Black': 23296, 'Hispanic': 9022, 'Asian/Pacific Islander': 1326, 'Native American/Native Alaskan': 917}


# Reading In A Second Dataset
We explored gun deaths by race in the past screen. However, our analysis only gives us the total number of gun deaths by race in the US. Unless we know the proportion of each race in the US, we won't be able to meaningfully compare those numbers. What we really want to get is a rate of gun deaths per 100000 people of each race. In order to do this, we'll need to read in data about what percentage of the US population falls into each racial category. Luckily, we can import some census data to help us out.

The data contains information on the total population of the US, as well as the total population of each racial group in the US. The data is stored in the census.csv file, and only consists of two rows:

We'll need to read this file in using the csv.reader() function.

Instructions
Read in census.csv, and convert to a list of lists. Assign the result to the census variable.
Display census to verify your work.

In [7]:
data2 = list(csv.reader(open('census.csv', 'r')))
print(data2)

mapping = {'White': 197318956, 
           'Native American/Native Alaskan' : 3739506, 
           'Hispanic' : 44618105, 
           'Black' : 40250635, 
           'Asian/Pacific Islander' : 15834141}

[['Id', 'Year', 'Id', 'Sex', 'Id', 'Hispanic Origin', 'Id', 'Id2', 'Geography', 'Total', 'Race Alone - White', 'Race Alone - Hispanic', 'Race Alone - Black or African American', 'Race Alone - American Indian and Alaska Native', 'Race Alone - Asian', 'Race Alone - Native Hawaiian and Other Pacific Islander', 'Two or More Races'], ['cen42010', 'April 1, 2010 Census', 'totsex', 'Both Sexes', 'tothisp', 'Total', '0100000US', '', 'United States', '308745538', '197318956', '44618105', '40250635', '3739506', '15159516', '674625', '6984195']]


# Computing Rates Of Gun Deaths Per Race

Earlier, we computed the number of gun deaths per race, and created a dictionary, race_counts, that looked like this:


{
     'Asian/Pacific Islander': 1326,
     'Black': 23296,
     'Hispanic': 9022,
     'Native American/Native Alaskan': 917,
     'White': 66237
}
In order to get from the raw counts of gun deaths by race to a rate of gun deaths per 100000 people in each race, we'll need to divide the total number of gun deaths by the population of each race and then multiply this by 100000. 

Here's a list of the race name in data, and the corresponding race name in census:

Asian/Pacific Islander -- Race Alone - Asian plus Race Alone - Native Hawaiian and Other Pacific Islander.
Black -- Race Alone - Black or African American.
Hispanic -- Race Alone - Hispanic
Native American/Native Alaskan -- Race Alone - American Indian and Alaska Native
White -- Race Alone - White
We'll need to create a dictionary that has each race name from data as a key, and has the population count for the races from census as the values.

Instructions
1. Manually create a dictionary, mapping that maps each key from race_counts to the population count of the race from census.
The keys in the dictionary should be Asian/Pacific Islander, Black, Native American/Native Alaskan, Hispanic, and White.
In the case of Asian/Pacific Islander, you'll need to add the counts from census for Race Alone - Asian, and Race Alone - Native Hawaiian and Other Pacific Islander.
2. Create an empty dictionary, race_per_hundredk.
3. Loop through each key in race_counts.
Divide the value associated with the key in race_counts by the value associated with the key in mapping.
Multiply by 100000.
Assign the result to the same key in race_per_hundredk.
4. When you're done, race_per_hundredk should contain the rate of gun deaths per 100000 people for each racial category.
5. Print race_per_hundredk to verify your work.

In [11]:
race_per_hundredk = {}

for i in race_counts:
    race_per_hundredk[i] = (race_counts[i]/mapping[i])*100000

In [12]:
race_per_hundredk

{'Asian/Pacific Islander': 8.374309664161762,
 'Black': 57.8773477735196,
 'Hispanic': 20.220491210910907,
 'Native American/Native Alaskan': 24.521955573811088,
 'White': 33.56849303419181}

# Filtering By Intent
We can filter our results, and restrict them to the Homicide intent. This will tell us what the gun-related murder rate per 100000 people in each racial category is. In order to do this, we'll need to redo our work in generating race_counts, but only count rows where the intent was Homicide.

Instructions
1. Extract the intent column using a list comprehension. The intent column is the fourth column in data.
Assign the result to intents.
2. Extract the race column using a list comprehension. The race column is the eighth column in data.
Assign the result to races.
3. Create an empty dictionary called homicide_race_per_hundredk
4. Use the enumerate() function to loop through each item in races. The position should be assigned to the loop variable i, and the value to the loop variable race.
Check the value at position i in intents.
If the value at position i in intents is Homicide:
If the key race doesn't exist in homicide_race_per_hundredk, create it.
Add 1 to the value associated with race in homicide_race_per_hundredk.
5. When you're done, homicide_race_per_hundredk should have one key for each of the racial categories in data. The associated value should be the number of gun deaths by homicide for that race.
6. Perform the same procedure we did in the last screen using mapping on homicide_race_per_hundredk to get from raw numbers to rates per 100000.
7. Display homicide_race_per_hundredk to verify your work.
8. Write up your findings in a markdown cell.
9. Write up any next steps you want to pursue with the data in a markdown cell.

In [13]:
intents = [i[3] for i in data]
races = [i[7] for i in data]

homicide_race_per_hundredk = {}

for i, j in enumerate(races):
    if intents[i] == 'Homicide':
        if j in homicide_race_per_hundredk:
            homicide_race_per_hundredk[j] = homicide_race_per_hundredk[j] + 1
        else:
            homicide_race_per_hundredk[j] = 1

In [14]:
homicide_race_per_hundredk


{'Asian/Pacific Islander': 559,
 'Black': 19510,
 'Hispanic': 5634,
 'Native American/Native Alaskan': 326,
 'White': 9147}

In [15]:
for i,j in homicide_race_per_hundredk.items():
    homicide_race_per_hundredk[i] = (j/mapping[i])*100000
    

In [16]:
homicide_race_per_hundredk

{'Asian/Pacific Islander': 3.530346230970155,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}